### PR TITLE
[JENKINS-68699] Upgrade Yarn from 1.22.19 to 3.2.0

### PR DESCRIPTION
**Pull request by StefanSpieker**
Upgrade to yarn 3.0 and minor node update. Accoring to this [issue](https://github.com/eirslett/frontend-maven-plugin/issues/928) we still need to install the latest 1.x release to use the 3.0 ...

### Proposed changelog entries

* Internal: Upgrade to yarn 3.0

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/5824"></a>

### CODE REVIEW

File: war/.yarnrc.yml
Comment (timja) [reply war/.yarnrc.yml:1]: I've done a quick test and it looks like not required and it's preferable not to use this, but then we need to decide if we are using the zero-install mode or not.

It appears you've setup the git ignore file for the zero-install mode but it's not working because there's a `*.zip` ignore rule somewhere preventing it working.

if we got the zero install mode then we will commit 34mb of zip files containing dependencies.

and a 1.1mb `.pnp.cjs` file.

We would also need to add a `yarn install --check-cache` to CI to make sure no one tries to inject a dodgy dep.

We would also need to disable the `yarn install` in the pom file as it wouldn't be needed anymore.

Alternatively we can not use the zero install mode and then we should just add `.pnp.*` to the git ignore file.

Any thoughts?

Comment (timja) [reply war/.yarnrc.yml:1]: any opinion @StefanSpieker @fqueiruga @janfaracik ?

**Comment by StefanSpieker** [comment]
I think we should wait for https://github.com/eirslett/frontend-maven-plugin/pull/1012

**Comment by basil** [comment]
> I think we should wait for eirslett/frontend-maven-plugin#1012

@StefanSpieker eirslett/frontend-maven-plugin#1012 was released in [1.12.1](https://github.com/eirslett/frontend-maven-plugin/releases/tag/frontend-plugins-1.12.1), which we upgraded to in #6139.

**Comment by NotMyFault** [comment]
The outstanding blockers have been addressed by now, do you mind resolving the merge conflicts so we can pick the PR up for another review @StefanSpieker ?

**Comment by basil** [comment]
I resolved the hard/physical merge conflicts at least. There may still be soft/logical conflicts.

**Comment by basil** [comment]
I just pushed some commits that should complete this effort.

**Review by basil — DISMISSED** [review]

File: war/.yarnrc.yml
Comment (basil) [reply war/.yarnrc.yml:3]: > do we need node_modules?

I tried removing it but then `yarn build` just hung. I don't think this needs to be tackled right away. I suspect something in the ancient dependency tree in package.json is tripping it up.

**Comment by basil** [comment]
Oh dear, `war/.yarn/releases/yarn-3.2.0.cjs` is a 2.1 MiB file? And they want you to check this into your repository? I really dislike this. Not sure we want to do this. There has to be a better way.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
> Oh dear, `war/.yarn/releases/yarn-3.2.0.cjs` is a 2.1 MiB file? And they want you to check this into your repository? I really dislike this. Not sure we want to do this. There has to be a better way.

This might be what you are after: 
https://stackoverflow.com/a/71876960

**Comment by basil** [comment]
I can't make heads or tails out of that. If you can get that to work, push some commits to this repository. Otherwise I am -1 on checking in the huge file and plan to close this PR.

**Comment by timja** [comment]
Sure I'll take a look ~tonight

**Comment by basil** [comment]
Whatever this `corepack` thing is, it doesn't look like `frontend-maven-plugin` supports it, so I don't think there is a path forward here.

**Comment by basil** [comment]
Well I guess in the absence of explicit support from `frontend-maven-plugin`, I guess we could use `frontend-maven-plugin` to install Node and NPM, use NPM to install `corepack`, and use `corepack` to install Yarn. I am trying to think of a word to describe how I feel about that. Maybe "Rube Goldberg machine"...

**Comment by timja** [comment]
Yeah that was my conclusion too, I got it working outside of maven but not through the front end maven plugin

**Comment by basil** [comment]
My personal feeling is that we should let some other party be the beta testers for the latest Yarn releases so I am proposing this for closer.

**Comment by timja** [comment]
> I suppose we could at least offer `maven-frontend-plugin` a feature request, if not a PR. Do you recall the steps you followed to get this working ountside of `maven-frontend-plugin`?

Roughly"

- Add `"packageManager": "yarn@3.2.0",` to `war/package.json`
- Delete `yarnPath` from the yarnrc config file.
- Remove checked in yarn v3 from the repo
- Run `corepack enable`
- Ensure corepack shims are ahead of any existing yarn in the path (I just deleted mine for expediency)
- Run `yarn install`

No yarn v3 appeared in the repo and it was running with yarn v3.
I assume there's some cache folder it bootstraps it to somewhere, didn't check where though.

**Comment by basil** [comment]
To summarize, this PR is functionally complete, but it requires checking in a 2.1 MiB Yarn Berry binary blob into the Git repository, which we would rather not do. This requirement is not a limitation of Yarn; rather, it is a limitation of the current release of `frontend-maven-plugin`, which lacks support for Corepack (which is the official way to manage package managers without checking in binary blobs). Improving `frontend-maven-plugin` to support Corepack is tracked in https://github.com/eirslett/frontend-maven-plugin/issues/1032, and this effort is significantly outside of the scope of the Jenkins project. Therefore I am closing this PR due to both inactivity on the part of the original author as well as the lack of feasibility of the current implementation.

---

On behalf of the core team, I would like to thank you for your contribution. Even though this PR did not make it across the finish line, it was a promising start! I continue to encourage you (or anyone else who is interested!) to pick up this effort and drive it to completion. Thanks!

**Comment by basil** [comment]
In commit b54f044b6ca047b40883a028586b9e7315cff1e6 I found a trick to download Yarn Berry via Maven Antrun Plugin rather than Frontend Maven Plugin. If accepted, it is _critical_ that this PR be squash merged to avoid bloating the repository size.

**Review by timja — APPROVED** [review]
I tested it out locally and all seems good

**Review by NotMyFault — APPROVED** [review]
Thanks for investigating into an alternative solution, basil, works out great so far when checking out locally.

---

Non blocking nit: berry 3.2.1 is out by now which we could use over 3.2.0.

**Comment by basil** [comment]
> Non blocking nit

I am glad this is non-blocking, because I would decline to accommodate it after going above and beyond the expectations for a maintainer to the degree that I already have regarding this PR which I did not submit myself.

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: war/.yarnrc.yml
Comment (timja) [reply war/.yarnrc.yml:3]: Seems to work fine on my machine, I'll file a PR after this is merged

**Comment by daniel-beck** [comment]
Renovate doesn't seem to like this in #6892.

**Comment by timja** [comment]
> Renovate doesn't seem to like this in #6892.

I'll take a look, renovate definitely supports it

**Comment by timja** [comment]
looking at https://github.com/renovatebot/renovate/issues/14797

**Comment by timja** [comment]
see https://github.com/renovatebot/renovate/pull/15414

**Comment by basil** [comment]
I pushed commit 33ba459242d9d91ed84f472e8fdbd4f1a295f9e4 to  #6894 to make the hack even more gross to be able to support Renovate.

**Comment by Vlatombe** [comment]
Not sure this matters, but the antrun execution breaks offline usage.

**Comment by basil** [comment]
The antrun execution is a hack to work around https://github.com/eirslett/frontend-maven-plugin/issues/1032. The long-term solution should be implemented in `frontend-maven-plugin` and would support offline usage.

**Pull request merged**
